### PR TITLE
[SPARK-15359] [Mesos] Mesos dispatcher should handle DRIVER_ABORTED status from mesosDriver.run()

### PR DESCRIPTION
**Pull request by devaraj-kavali**
## What changes were proposed in this pull request?

When the mesosDriver.run() returns with the status as DRIVER_ABORTED then throwing the exception which can be handled from SparkUncaughtExceptionHandler to shutdown the dispatcher.
## How was this patch tested?

I verified it manually, the driver thread throws exception when mesosDriver.run() returns with the DRIVER_ABORTED status.

### CODE REVIEW

**Comment by tnachen** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala [hunk 1]
```diff
@@ -120,14 +120,25 @@ private[mesos] trait MesosSchedulerUtils extends Logging {
             val ret = mesosDriver.run()
             logInfo("driver.run() returned with code " + ret)
             if (ret != null && ret.equals(Status.DRIVER_ABORTED)) {
-              error = Some(new SparkException("Error starting driver, DRIVER_ABORTED"))
-              markErr()
+              val ex = new SparkException("Error starting driver, DRIVER_ABORTED")
+              // if the driver gets aborted after the successful registration
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala]: s/after the successful registration/after registration/g

**Comment by tnachen** [comment]
Is it because MesosDriver actually threw an exception?

**Comment by devaraj-kavali** [comment]
MesosDriver doesn't throw any exception, it just returns with the value as Status.DRIVER_ABORTED.

     ```
      registerLatch.await()

      // propagate any error to the calling thread. This ensures that SparkContext creation fails
      // without leaving a broken context that won't be able to schedule any tasks
      error.foreach(throw _)

This code handles exceptions and throws if it gets Status.DRIVER_ABORTED during registration, once the registration completes there is no code to handle and will be skipped the status and also thread dies.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala (hunk 1 above)
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala]: Also to simplify the code, can we just throw SparkExecption here? Then the catch will then handle all cases

**Comment by devaraj-kavali** [comment]
@tnachen Can you check this?

**Comment by devaraj-kavali** [comment]
@mgummelt /@tnachen, can you have a look into this?

**Comment by mgummelt** [comment]
What whole function is designed poorly.  We need to totally change it instead of tacking this on.  We shouldn't be calling `driver.run()` in a separate thread.  We should be calling `driver.start()`

**Comment by devaraj-kavali** [comment]
Thanks @mgummelt for the feedback, will update the PR with the function rewrite.

**Comment by skonto** [comment]
@ArtRand @susanxhuynh could please review this before we call for a merge?

**Comment by ArtRand** [comment]
@skonto @susanxhuynh @devaraj-kavali are people still interested in this? I was just playing around with this code to clean up ZK state.. Would be happy to try this when I have a few cycles.

**Comment by devaraj-kavali** [comment]
@ArtRand I think this is still an issue which needs to be merged, do you have any observations with this PR?

**Comment by ArtRand** [comment]
Hello @devaraj-kavali. Yes. I've been playing around with this because it's inconvenient to clean up ZK whenever you uninstall/reinstall the Dispatcher. The problem is that the only signal of a re-install vs. a failover is when Mesos gives you a `DRIVER_ABORTED` error (signals a re-install). I think a solution will involve having different methods to starting `SchedulerDrivers` for Spark Drivers (applications) and the Dispatcher because they inherently have different requirements with respect to state.

**Comment by HyukjinKwon** [comment]
ping @devaraj-kavali for @ArtRand's comment above.

**Pull request closed without merging**